presentation-compiler/src/main/dotty/tools/pc/AutoImports.scala

```diff
@@ -0,0 +1,378 @@
+                  )
+
+              SymbolImport(
+                symbol,
+                SymbolIdent.Select(
+                  SymbolIdent.direct(rename),
+                  symbol.nameBackticked(false)
+                ),
+                importSel
+              )
+            case None =>
+              SymbolImport(
+                symbol,
+                SymbolIdent.direct(symbol.fullNameBackticked),
+                None
+              )
+          end match
+        case IndexedContext.Result.InScope =>
+          val direct = renames(symbol).getOrElse(symbol.nameBackticked)
+          SymbolImport(symbol, SymbolIdent.direct(direct), None)
+      end match
+    end inferSymbolImport
+
+    def renderImports(
+        imports: List[ImportSel]
+    )(using Context): Option[l.TextEdit] =
+      if imports.nonEmpty then
+        val indent0 = " " * importPosition.indent
+        val editPos = pos.withSpan(Spans.Span(importPosition.offset)).toLsp
+
+        // for worksheets, we need to remove 2 whitespaces, because it ends up being wrapped in an object
+        // see WorksheetProvider.worksheetScala3AdjustmentsForPC
+        val indent =
+          if pos.source.path.isWorksheet &&
+            editPos.getStart().getCharacter() == 0
+          then indent0.drop(2)
+          else indent0
+        val topPadding =
+          if importPosition.padTop then "\n"
+          else ""
+
+        val formatted = imports
+          .map {
+            case ImportSel.Direct(sym) => importName(sym)
+            case ImportSel.Rename(sym, rename) =>
+              s"${importName(sym.owner)}.{${sym.nameBackticked(false)} => $rename}"
+          }
+          .map(sel => s"${indent}import $sel")
+          .mkString(topPadding, "\n", "\n")
+
+        Some(new l.TextEdit(editPos, formatted))
+      else None
+    end renderImports
+
+    private def importName(sym: Symbol): String =
+      if indexedContext.importContext.toplevelClashes(sym) then
+        s"_root_.${sym.fullNameBackticked(false)}"
+      else sym.fullNameBackticked(false)
+  end AutoImportsGenerator
+
+  private def autoImportPosition(
+      pos: SourcePosition,
+      text: String,
+      tree: Tree,
+      comments: List[Comment]
+  )(using Context): AutoImportPosition =
+
+    @tailrec
+    def lastPackageDef(
+        prev: Option[PackageDef],
+        tree: Tree
+    ): Option[PackageDef] =
+      tree match
+        case curr @ PackageDef(_, (next: PackageDef) :: Nil)
+            if !curr.symbol.isPackageObject =>
+          lastPackageDef(Some(curr), next)
+        case pkg: PackageDef if !pkg.symbol.isPackageObject => Some(pkg)
+        case _ => prev
+
+    def firstObjectBody(tree: Tree)(using Context): Option[Template] =
+      tree match
+        case PackageDef(_, stats) =>
+          stats.flatMap {
+            case s: PackageDef => firstObjectBody(s)
+            case TypeDef(_, t @ Template(defDef, _, _, _))
+                if defDef.symbol.showName == "<init>" =>
```

Comment:
```suggestion
                if defDef.symbol.isConstructor =>
```
In general we should only use show when pretty-printing, not inside logical checks.